src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -698,4 +577,266 @@ public SNIError GetLastError()
+
+            if (_dataSourceAfterTrimmingProtocol.Contains("/")) // Pipe paths only allow back slashes
+            {
+                if (ConnectionProtocol == DataSource.Protocol.None)
+                    ReportSNIError(SNIProviders.INVALID_PROV);
+                else if (ConnectionProtocol == DataSource.Protocol.NP)
+                    ReportSNIError(SNIProviders.NP_PROV);
+                else if (ConnectionProtocol == DataSource.Protocol.TCP)
+                    ReportSNIError(SNIProviders.TCP_PROV);
+            }
+        }
+
+        private void populateProtocol()
+        {
+            string[] splitByColon = _workingDataSource.Split(':');
+
+            if (!(splitByColon.Length > 1))
+            {
+                ConnectionProtocol = DataSource.Protocol.None;
+            }
+            else
+            {
+                // We trim before switching because " tcp : server , 1433 " is a valid data source
+                switch (splitByColon[0].Trim())
+                {
+                    case TdsEnums.TCP:
+                        ConnectionProtocol = DataSource.Protocol.TCP;
+                        break;
+                    case TdsEnums.NP:
+                        ConnectionProtocol = DataSource.Protocol.NP;
+                        break;
+                    case TdsEnums.ADMIN:
+                        ConnectionProtocol = DataSource.Protocol.Admin;
+                        break;
+                    default:
+                        // None of the supported protocols were found. This may be a IPv6 address
+                        ConnectionProtocol = DataSource.Protocol.None;
+                        break;
+                }
+            }
+        }
+
+        public static DataSource ParseServerName(string dataSource)
+        {
+            DataSource details = new DataSource(dataSource);
+
+            if (details.IsBadDataSource)
+            {
+                return null;
+            }
+
+            if (details.InferNamedPipesInformation() && !details.IsBadDataSource)
+            {
+                return details;
+            }
+            
+            if(details.IsBadDataSource)
+            { 
+                return null;
+            }
+
+            if (!details.InferConnectionDetails() && details.IsBadDataSource)
+            {
+                return null;
+            }
+
+            return details;
+        }
+
+        private bool InferLocalServerName()
+        {
+            // If Server name is empty or localhost, then use "localhost"
+            if (string.IsNullOrEmpty(ServerName) || IsLocalHost(ServerName))
+            {
+                if (ConnectionProtocol == DataSource.Protocol.Admin)
+                {
+                    ServerName = Environment.MachineName;
+                }
+                else
+                {
+                    ServerName = DefaultHostName;
+                }
```

Comment:
Nit: personal preference, but I like:
```C#
ServerName = ConnectionProtocol == DataSource.Protocol.Admin ?
    Environment.MachineName :
    DefaultHostName;
```
as it helps to highlight what the branch is actually affecting by minimizing the differences between them.